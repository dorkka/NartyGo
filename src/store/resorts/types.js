export const SET_RESORTS = 'resorts/SET_RESORTS';
export const IS_LOADING = 'resorts/IS_LOADING';
export const SET_ERROR = 'resorts/SET_ERROR';
export const SET_RESORT = 'resorts/SET_RESORT';
export const SET_SELECTED_OPTIONS = 'SET_SELECTED_OPTIONS';
