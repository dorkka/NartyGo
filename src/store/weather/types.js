export const SET_WEATHER = 'weather/SET_WEATHER';
export const IS_LOADING = 'weather/IS_LOADING';
export const SET_ERROR = 'weather/SET_ERROR';
