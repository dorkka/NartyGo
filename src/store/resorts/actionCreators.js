import * as types from './types';
import resourceFetcher from '../../services/resourceFetcher';

export const setResorts = (data, headers, perPage) => {
  const pageCount = Math.ceil(headers.get('x-total-count') / perPage);
  const list = [];
  const byId = {};
  data.forEach(element => {
    list.push(element.id);
    byId[element.id] = element;
  });
  return {
    type: types.SET_RESORTS,
    payload: {
      pageCount,
      list,
      byId,
    },
  };
};

export const setError = (error) => ({
  type: types.SET_ERROR,
  error,
});

export const setIsLoading = () => ({
  type: types.IS_LOADING,
});

export const setSpecificResort = data => {
  const byId = { [data.id]: data };
  return {
    type: types.SET_RESORT,
    payload: {
      byId,
    },
  };
};

export const getResorts = (_page, _limit) => (dispatch) => {
  dispatch(setIsLoading());
  return resourceFetcher('http://localhost:3001/resorts')({ _page, _limit })
    .then(({ data, headers }) => {
      dispatch(setResorts(data, headers, _limit));
    })
    .catch(error => dispatch(setError(error)));
};

export const getSpecificResort = (id) => (dispatch, getState) => {
  if ((getState().resorts.byId[id])) {
    return null;
  }
  dispatch(setIsLoading());
  return resourceFetcher(`http://localhost:3001/resorts/${id}`)()
    .then(({ data }) => {
      dispatch(setSpecificResort(data));
    })
    .catch(error => dispatch(setError(error)));
};

export const setSelectOptions = (data) => {
  const resortsOptions = []
  const cityOptions = []
  let label = ''
  data.forEach((element) => {
    label = `${element.name} - ${element.city}`;
    resortsOptions.push({ value: element.name, label, id: element.id })
    cityOptions.push({ value: element.city, label, id: element.id })
  })
  return {
    type: types.SET_SELECTED_OPTIONS,
    payload: {
      resortsOptions,
      cityOptions,
    }
  }
}

export const getAllResorts = () => (dispatch) => {
  resourceFetcher('http://localhost:3001/resorts')().then(({ data }) => {
    dispatch(setSelectOptions(data))
  });
}
