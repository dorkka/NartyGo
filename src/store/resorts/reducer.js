import * as types from './types';

const initialState = {
  byId: {},
  list: [],
  pageCount: 1,
  isLoading: false,
  error: null,
};

function resortsReducer(state = initialState, action) {
  switch (action.type) {
    case types.SET_RESORTS: {
      return {
        ...state,
        byId: {
          ...state.byId,
          ...action.payload.byId,
        },
        list: action.payload.list,
        pageCount: action.payload.pageCount,
        isLoading: false,
      };
    }
    case types.SET_ERROR: {
      return { ...state, error: action.payload.error };
    }
    case types.IS_LOADING: {
      return { ...state, isLoading: true };
    }
    default: {
      return state;
    }
  }
}
export default resortsReducer;
