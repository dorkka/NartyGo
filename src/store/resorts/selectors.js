export const getCurrentResorts = (state) =>
  state.resorts.list.map((id) => state.resorts.byId[id] || {});
export const getSpecificResort = (state, id) =>
  state.resorts.byId[id] || { weather: {}, piste: [] };
