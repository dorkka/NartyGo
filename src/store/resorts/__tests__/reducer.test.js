import resortsReducer from '../reducer';
import * as types from '../types';
import { resorts } from '../../../specs/fixtures/resorts';

describe('resortsReducer', () => {
  const initialState = {
    byId: {},
    list: [],
    pageCount: 1,
    isLoading: false,
    error: null,
  };

  test('should return inital state when resortsReducer is called with default params', () => {
    expect(resortsReducer(undefined, {})).toEqual(initialState);
  });

  // using snapshot
  test('should return appropriate state when action type is SET_RESORTS', () => {
    const byId = {
      [resorts[0].id]: resorts[0],
      [resorts[1].id]: resorts[1],
    };
    const action = {
      type: types.SET_RESORTS,
      payload: {
        pageCount: 5,
        list: [1, 2],
        byId,
      },
    };
    expect(resortsReducer(initialState, action)).toMatchSnapshot();
  });

  // using toEqual();
  test('should return appropriate state when action type is SET_RESORT', () => {
    const byId = {
      [resorts[0].id]: resorts[0],
    };
    const action = {
      type: types.SET_RESORT,
      payload: {
        byId,
      },
    };
    const newState = {
      byId,
      list: [],
      pageCount: 1,
      isLoading: false,
      error: null,
    };
    expect(resortsReducer(initialState, action)).toEqual(newState);
  });

  test('should return appropriate state when action type is SET_ERROR', () => {
    expect(resortsReducer(initialState, { type: types.SET_ERROR, error: { mesage: 'test error' } })).toMatchSnapshot();
  });

  test('should return state with truthy isLoading when action type is IS_LOADING', () => {
    expect(resortsReducer(initialState, { type: types.IS_LOADING })).toMatchSnapshot();
  });
});
