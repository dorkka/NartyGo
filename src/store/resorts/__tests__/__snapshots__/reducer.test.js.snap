// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`resortsReducer should return appropriate state when action type is SET_ERROR 1`] = `
Object {
  "byId": Object {},
  "error": Object {
    "mesage": "test error",
  },
  "isLoading": false,
  "list": Array [],
  "pageCount": 1,
}
`;

exports[`resortsReducer should return appropriate state when action type is SET_RESORTS 1`] = `
Object {
  "byId": Object {
    "1": Object {
      "city": "Ustroń",
      "cityId": 7531846,
      "coordinates": Object {
        "lat": 49.6833162,
        "lng": 18.8196089,
      },
      "id": 1,
      "name": "Czantoria",
      "piste": Array [
        Object {
          "id": 1,
          "name": "red",
          "number": 1,
        },
        Object {
          "id": 2,
          "name": "blue",
          "number": 2,
        },
      ],
    },
    "2": Object {
      "city": "Karlików",
      "cityId": 7531370,
      "coordinates": Object {
        "lat": 49.43529,
        "lng": 22.067767,
      },
      "id": 2,
      "name": "Karlików",
      "piste": Array [
        Object {
          "id": 1,
          "name": "Mała Palenica",
          "number": 1,
        },
      ],
    },
  },
  "error": null,
  "isLoading": false,
  "list": Array [
    1,
    2,
  ],
  "pageCount": 5,
}
`;

exports[`resortsReducer should return state with truthy isLoading when action type is IS_LOADING 1`] = `
Object {
  "byId": Object {},
  "error": null,
  "isLoading": true,
  "list": Array [],
  "pageCount": 1,
}
`;
